### ISSUE TITLE
Dealing with special keys

### ISSUE THREAD

**Issue by mgedmin** [dohsimpson/vim-macroeditor#1]
When you record a vim macro, any special keys (such as arrows, or Backspace) get encoded using these strange `<80>kX` sequences.  Do you think it would be possible for a plugin like yours to convert them to readable key names such as `<Up>` etc. and then convert them back after editing?

**Comment by dohsimpson** [comment]
Thanks for the suggestion, but I don't see an easy way to accomplish it. Ideally if there a Vim function that does that, it would be easy to implement.

**Comment by juanMarinero** [comment]
I opened same issue in another trending Vim plugin: [link](https://github.com/junegunn/vim-peekaboo/issues/78). No much hope though.

**Comment by llakala** [comment]
Seven years later, I have a solution! [`keytrans()`](https://vimhelp.org/builtin.txt.html#keytrans%28%29) will translate any control characters to their normal mapping equivalents - so `^M` gets turned into `<CR>`.

However, once you go in one direction, you need to reverse the process - so you can write a macro as if it's a normal keymap, and convert it in the background to use control characters. Unfortunately, I wasn't able to find a vimscript function to do this. But I *did* find `nvim_replace_termcodes`, a Neovim function[^1] that does exactly what we're looking for.

This is the relevant patch to make the plugin convert from/to special characters: 

```patch
diff --git a/plugin/macroeditor.vim b/plugin/macroeditor.vim
index 6e1eb98..ddc4776 100644
--- a/plugin/macroeditor.vim
+++ b/plugin/macroeditor.vim
@@ -1,5 +1,6 @@
 function! YankToRegister()
-  exe printf('norm! ^"%sy$', b:registername)
+  let contents = getline(".")
+  call setreg(b:registername, v:lua.vim.api.nvim_replace_termcodes(contents, 1, 0, 1))
 endfunction
 
 function! OpenMacroEditorWindow(registername)
@@ -16,7 +17,8 @@ function! OpenMacroEditorWindow(registername)
   execute height 'new ' name
   let b:registername = a:registername
   setlocal bufhidden=wipe noswapfile nobuflisted
-  exe printf('norm! "%sp', b:registername)
+  let contents = keytrans(getreg(b:registername))
+  exe setline(".", contents)
   set nomodified
   augroup MacroEditor
     au!
```

I also put the commit up at https://github.com/llakala/vim-macroeditor. Neovim-only, for obvious reasons.

[^1]: If anyone knows a way to do this with vimscript only, let me know!

**Comment by juanMarinero** [comment]
@llakala I quick tested it, it work! **Thanks**!

Just some minor issues:
- `nvim_replace_termcodes` documentation [link](https://neovim.io/doc/user/api.html#nvim_replace_termcodes())
- The plugin is in your fork, thus it's installed via `Plug 'llakala/vim-macroeditor'` (or alike). Not with ~~`Plug 'dohsimpson/vim-macroeditor'`~~. Alternative use `Plug 'tuurep/registereditor'` from [https://github.com/tuurep/registereditor](https://github.com/tuurep/registereditor)
- Maybe the original plugin owner would gladly accept a PR? What you say @dohsimpson ?
- Maybe the PR could still be used in plain Vim (not NeoVim) adding some `if`-s, like:

```vim
if has('nvim')
  let contents = getline(".")
  call setreg([...]
else
  exe printf('norm! ^"%sy$', b:registername)
endif

if has('nvim')
  let contents = keytrans(getreg(b:registername))
  exe setline(".", contents)
else
  exe printf('norm! "%sp', b:registername)
endif
```

**Comment by juanMarinero** [comment]
@llakala I think you can promote the fork [tuurep/registereditor](https://github.com/tuurep/registereditor) here, in this issue comments, because I guess you first solved it in [https://github.com/tuurep/registereditor/issues/4](https://github.com/tuurep/registereditor/issues/4) and later you applied that approach in current Vim plugin (well, in https://github.com/llakala/vim-macroeditor). Thanks for **solving current issue**, and not just letting it unsolved. Excellent commitment Sir!

@dohsimpson I think that [tuurep/registereditor](https://github.com/tuurep/registereditor) is quite expanding your original awesome plugin. Thus, maybe you could add at top of your README a quick note pointing to this **Neovim-only** fork please.

@dohsimpson I also suggest a **disclaimer** in the README if no further maintaining this repo.

**Comment by dohsimpson** [comment]
@juanMarinero thanks for letting me know! Yes, let me update the README.

**Comment by llakala** [comment]
> I think you can promote the fork [tuurep/registereditor](https://github.com/tuurep/registereditor) here, in this issue comments, because I guess you first solved it in https://github.com/tuurep/registereditor/issues/4 and later you applied that approach in current Vim plugin (well, in [llakala/vim-macroeditor](https://github.com/llakala/vim-macroeditor)).

Yep, feel free to access it at https://github.com/llakala/vim-macroeditor. Note that it's meant to be fairly evergreen - I don't plan on doing heavy maintenance. If you want something that's actually maintained, go to https://github.com/tuurep/registereditor.

**Comment by tuurep** [comment]
Thanks for all the support everyone, much appreciated!
